### PR TITLE
Static features and callable/catalogue support

### PR DESCRIPTION
**Pull request by martbern**
Primarily `add_age` and `add_static_predictor` in `flattened_dataset.py`.

Tests are `test_static_predictor` and `test_add_age` in `test_add_values.py`.

I've also renamed a bunch of methods to separate static from temporal features.

### CODE REVIEW

**Comment by martbern** [comment]
Hmm, looks like the dev branch wasn't properly merged, sorry! Only review changes in the last commit (cbeedec).

**Comment by martbern** [comment]
I've tried adding catalogue support in c08d53d, but for some reasons, it doesn't register `load_event_times` in the catalogue even though I explicitly import it in `test_add_values.py`. @KennethEnevoldsen, do you know what's going on here?

**Comment by martbern** [comment]
Hmm, so it appears the loaders registry is emptied when calling pool.map in line 166 of `flattened_dataset` :-(

**Comment by martbern** [comment]
Okay, seems to fix it if I resolve the catalogue to a dict before mapping! 

Sorry about the multiple bits of functionality in the same PR.

File: src/psycopmlutils/timeseriesflattener/flattened_dataset.py [hunk 1]
```diff
@@ -73,18 +76,19 @@ def __init__(
...
         self,
         predictor_list: List[Dict[str, str]],
-        predictor_dfs_dict: Dict[str, DataFrame],
+        predictor_dfs_dict: Dict[str, DataFrame] = None,
         resolve_multiple_fn_dict: Optional[Dict[str, Callable]] = None,
     ):
         """Add predictors to the flattened dataframe from a list.
 
         Args:
             predictor_list (List[Dict[str, str]]): A list of dictionaries describing the prediction_features you'd like to generate.
```
Comment (KennethEnevoldsen) [inline src/psycopmlutils/timeseriesflattener/flattened_dataset.py:90]: remove type hint for argument name
```suggestion
            predictors (List[Dict[str, str]]): A list of dictionaries describing the prediction_features you'd like to generate.
```

File: src/psycopmlutils/timeseriesflattener/flattened_dataset.py [hunk 2]
```diff
@@ -73,18 +76,19 @@ def __init__(
...
-            predictor_dfs (Dict[str, DataFrame]): A dictionary mapping the predictor_df in predictor_list to DataFrame objects.
+            predictor_dfs_dict (Dict[str, DataFrame], optional): A dictionary mapping the predictor_df in predictor_list to DataFrame objects. Optional, can be replaced by adding callables returning dfs to @data_loaders. Defaults to None.
```
Comment (KennethEnevoldsen) [inline src/psycopmlutils/timeseriesflattener/flattened_dataset.py:91]: remove type hint and split line.

Type hint does not allow for None, either update type hint or change to {}
```suggestion
            predictor_dfs (Dict[str, DataFrame], optional): A dictionary mapping the predictor_df in predictor_list to DataFrame objects. Defaults to None, in which case it uses [...]. Can be replaced by adding callables  returning dfs to @data_loaders. 
```
Not sure what is meant by:
```
Can be replaced by adding callables  returning dfs to @data_loaders. Defaults to None.
```
Generally not sure what the default is.

File: src/psycopmlutils/timeseriesflattener/flattened_dataset.py [hunk 3]
```diff
@@ -178,24 +190,66 @@ def add_predictors_from_list_of_argument_dictionaries(
...
             prediction_times_with_uuid_df=self.pred_times_with_uuid,
             id_col_name=self.id_col_name,
             timestamp_col_name=self.timestamp_col_name,
             pred_time_uuid_col_name=self.pred_time_uuid_col_name,
             **kwargs_dict,
         )
 
-    def add_outcome(
+    def add_age(
+        self, date_of_birth_df: DataFrame, date_of_birth_col_name: str = "date_of_birth"
```
Comment (KennethEnevoldsen) [inline src/psycopmlutils/timeseriesflattener/flattened_dataset.py:211]: Add docstring. Why does the date_of_birth_df contai? I assume it is a mapping in which case I would use date_of_birth_mapping.

File: src/psycopmlutils/timeseriesflattener/flattened_dataset.py [hunk 4]
```diff
@@ -178,24 +190,66 @@ def add_predictors_from_list_of_argument_dictionaries(
...
+            / (365.25)
+        ).round(2)
+
+        self.df_aggregating.drop(date_of_birth_col_name, axis=1, inplace=True)
+        self.df.drop(date_of_birth_col_name, axis=1, inplace=True)
+
+        self.df_aggregating["age_in_years"] = age
+        self.df["age_in_years"] = age
+
+    def add_static_predictor(self, predictor_df: DataFrame):
```
Comment (KennethEnevoldsen) [inline src/psycopmlutils/timeseriesflattener/flattened_dataset.py:239]: add docsting. What does the predictor df contain?

File: src/psycopmlutils/timeseriesflattener/flattened_dataset.py [hunk 5]
```diff
@@ -321,22 +375,31 @@ def create_flattened_df_for_value(
...
         Args:
-            values_df (DataFrame): A table in wide format. Required columns: patient_id, timestamp, value.
+            values_df_name (str): The name of the values_df to look for in the data_loaders catalogue or the values_df_dict.
             direction (str): Whether to look "ahead" or "behind".
             interval_days (float): How far to look in direction.
             resolve_multiple (Callable, str): How to handle multiple values within interval_days. Takes either i) a function that takes a list as an argument and returns a float, or ii) a str mapping to a callable from the resolve_multiple_fn catalogue.
             fallback (List[str]): What to do if no value within the lookahead.
             new_col_name (str, optional): Name to use for new column. Automatically generated as '{new_col_name}_within_{lookahead_days}_days'.
             source_values_col_name (str, optional): Column name of the values column in values_df. Defaults to "val".
+            values_df_dict (Dict[str, DataFrame], optional): _description_. Defaults to None.
```
Comment (KennethEnevoldsen) [inline src/psycopmlutils/timeseriesflattener/flattened_dataset.py:390]: Update docsting.

File: src/psycopmlutils/timeseriesflattener/flattened_dataset.py [hunk 6]
```diff
@@ -73,18 +76,19 @@ def __init__(
         # in self.df
         self.df_aggregating = self.pred_times_with_uuid
         self.df = self.pred_times_with_uuid
```
Comment (HLasse) [inline src/psycopmlutils/timeseriesflattener/flattened_dataset.py:78]: Depending on how these are used later on, might run into issues of shared memory, i.e. modifying `self.pred_times_with_uuid` might also modify `self.df` and vice versa. https://stackoverflow.com/questions/27673231/why-should-i-make-a-copy-of-a-data-frame-in-pandas

**Review by HLasse — COMMENTED** [review]
LGTM!

File: src/psycopmlutils/timeseriesflattener/flattened_dataset.py (hunk 6 above)
Comment (martbern) [reply src/psycopmlutils/timeseriesflattener/flattened_dataset.py:78]: Ah, that's excellent, thanks! Have added a .copy() where relevant.

**Pull request merged**